src/async_traits.rs

```diff
@@ -48,40 +64,175 @@ where
+        f: Func,
+        retry: RetryFunc,
+    ) -> Result<R, DieselError>
+    where
+        R: Send + 'static,
+        Fut: Future<Output = Result<R, DieselError>> + Send,
+        Func: Fn(SingleConnection<Conn>) -> Fut + Send + Sync,
+        RetryFut: Future<Output = bool> + Send,
+        RetryFunc: Fn() -> RetryFut + Send + Sync,
+    {
+        // Check out a connection once, and use it for the duration of the
+        // operation.
+        let conn = Arc::new(self.get_owned_connection());
+
+        // Refer to CockroachDB's guide on advanced client-side transaction
+        // retries for the full context here. In short, they expect a particular
+        // name for this savepoint, but Diesel has Opinions on savepoint names,
+        // so we use this session variable to identify that any name is valid.
+        //
+        // TODO: It may be preferable to set this once per connection -- but
+        // that'll require more interaction with how sessions with the database
+        // are constructed.
+        Self::start_transaction(&conn).await?;
+        conn.run_with_shared_connection(|conn| {
+            conn.batch_execute("SET LOCAL force_savepoint_restart = true")
+        })
+        .await?;
+
+        loop {
+            // Add a SAVEPOINT to which we can later return.
+            Self::add_retry_savepoint(&conn).await?;
+
+            let async_conn = SingleConnection(Self::as_async_conn(&conn).0.clone());
+            match f(async_conn).await {
+                Ok(value) => {
+                    // The user-level operation succeeded: try to commit the
+                    // transaction by RELEASE-ing the retry savepoint.
+                    if let Err(err) = Self::commit_transaction(&conn).await {
+                        // We're still in the transation, but we at least
+                        // tried to ROLLBACK to our savepoint.
+                        if !retryable_error(&err) || !retry().await {
+                            // Bail: ROLLBACK the initial BEGIN statement too.
+                            let _ = Self::rollback_transaction(&conn).await;
+                            return Err(err);
+                        }
+                        // ROLLBACK happened, we want to retry.
+                        continue;
+                    }
+
+                    Self::commit_transaction(&conn).await?;
+                    return Ok(value);
+                }
+                Err(user_error) => {
+                    // The user-level operation failed: ROLLBACK.
+                    if let Err(err) = Self::rollback_transaction(&conn).await {
+                        // If we fail while rolling back, prioritize returning
+                        // the ROLLBACK error over the user errors.
+                        return match Self::rollback_transaction(&conn).await {
+                            Ok(()) => Err(err),
+                            Err(err) => Err(err),
+                        };
```

Comment:
I'm not sure I follow this. We're rolling the transaction back twice. That seems like it should be "safe", but I don't understand why its needed. We're returning the `err` from the first call to `Self::rollback_transaction()` in any case, so why is there a second call?

Reply:
Ah, I didn't quite read this correctly. We're _not_ returning the error from the first call in any case, they're just named the same. I would recommend changing these to `user_err` or `transaction_err` and then `rollback_err` or similar, to disambiguate what we're actually doing here.

Also, what does the actual error on line 217 look like? Is that something like "can't rollback, not in transaction"? Or something else?

Reply:
Yeah, Diesel's implementation of `rollback_transaction` is basically just "peel back a single layer": https://docs.diesel.rs/master/src/diesel/connection/transaction_manager.rs.html#352-373 -- it issues `ROLLBACK TO SAVEPOINT diesel_savepoint_{}` if any are ongoing.

So when we're doing what the CRDB docs suggest -- BEGIN, SAVEPOINT XXX, ops, RELEASE SAVEPOINT XXX, COMMIT -- we have to issue two calls to `begin_transaction` on the incoming path, and two calls to `rollback_transaction` on the outgoing path.

I'll name these `first_rollback_error` and `second_rollback_error` to disambiguate.

> Also, what does the actual error on line 217 look like? Is that something like "can't rollback, not in transaction"? Or something else?

It's a https://docs.diesel.rs/master/diesel/result/enum.Error.html , so it **could** be anything here, but I think https://docs.diesel.rs/master/diesel/result/enum.Error.html#variant.RollbackErrorOnCommit seems possible.

Reply:
Ok, that makes sense, and the renaming sounds good. I might make a comment with the "two calls on the way in, two on the way out" idea, to make clear the first is rolling back to the savepoint, the second the entire transaction.